doc/developer/design/20220314_linear_or_eventual.md

```diff
@@ -0,0 +1,115 @@
+# Linearizable and eventually consistent reads
+
+## Summary
+
+Real-time sources come in two different flavors:
+- `materialized`, which means there is always an _index_ that drives reads for that source, and, critically, drives
+_timestamp bindings_, which are durably-stored associations between `timestamps` (the literal, real-time timestamps that mz uses to drive its dataflows)
+and `offsets` (consistent, ordered, source-specific integer identifiers that express how _far along_ you are in the source)
+- "unmaterialized", which have no such _index_. These sources could be queried, but produce inconsistent results that are not even _eventually consistent_.
+
+_timestamp bindings_ can allow for _linearizable_ reads, that is, reads that are consistent in respect to the offsets we read in real time. For example, a
+_linearizable_ read against a source will **always** include data for a recent offset, _and will associate a specific timestamp with that offset_.
+Later reads at a timestamp at or later than timestamp will _at least include_ data up to that offset.
+
+We want to fix the lack of any kind of consistency for "unmaterialized sources" AND offer _linearizable_ reads for all types of sources. In short, we want the following
+table to represent all possible interactions with sources in `materialized`:
+
+
+|              | unmaterialized | materialized   |
+| ------------ | -------------- | -------------- |
+| Enventual    | Not available  | Default        |
+| Linearizable | Default        | `LINEARIZABLE` |
+
+Not offering non-_linearizable_ reads for reads interacting with unmaterialized sources simplifies the implementation, and clarifies the user experience of
+working with unmaterialized sources
+
+## Goals
+
+1. Offer _linearizable_ reads in as many cases as possible. `materialized` is ultimately built to support this!
+2. Offer eventual consistency by default (keeping the same behavior we currently have in `materialized`) for materialized sources
+
+## Non-Goals
+
+Alter the default experience of using `materialized` (yet).
+
+## Description
+
+### Background
+
+Sources currently produce messages that contain offsets. Those offsets are written down, associated with a timestamp, as a _timestamp binding_.
```

Comment:
I think sources *consume* messages with offsets, and produce updates with timestamps.